File: experiments/adrianco/experiment-13/runs/language=go_model=sonnet_prompt=TDD/rep2/loader.go
package main

import (
	"encoding/csv"
	"io"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

// ParseDate tries multiple date formats and returns a parsed time.Time.
func ParseDate(s string) (time.Time, error) {
	formats := []string{
		"2006-01-02 15:04:05",
		"2006-01-02",
		"02/01/2006",
	}
	s = strings.TrimSpace(s)
	for _, fmt := range formats {
		t, err := time.Parse(fmt, s)
		if err == nil {
			return t, nil
		}
	}
	return time.Time{}, nil
}

func parseInt(s string) int {
	s = strings.TrimSpace(s)
	// Handle float strings like "1.0"
	if strings.Contains(s, ".") {
		f, err := strconv.ParseFloat(s, 64)
		if err != nil {
			return 0
		}
		return int(f)
	}
	n, err := strconv.Atoi(s)
	if err != nil {
		return 0
	}
	return n
}

func parseFloat(s string) float64 {
	s = strings.TrimSpace(s)
	f, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return 0
	}
	return f
}

// indexMap builds a map from header name -> column index (0-based).
func indexMap(headers []string) map[string]int {
	m := make(map[string]int, len(headers))
	for i, h := range headers {
		// Strip BOM from first header if present
		h = strings.TrimPrefix(h, "\xef\xbb\xbf")
		h = strings.TrimSpace(h)
		m[h] = i
	}
	return m
}

// LoadBrasileirao loads Brasileirao_Matches.csv
// Headers: "datetime","home_team","home_team_state","away_team","away_team_state","home_goal","away_goal","season","round"
func LoadBrasileirao(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	r := csv.NewReader(f)
	r.LazyQuotes = true
	headers, err := r.Read()
	if err != nil {
		return nil, err
	}
	idx := indexMap(headers)

	var matches []Match
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Printf("LoadBrasileirao: skipping row due to error: %v", err)
			continue
		}
		if len(row) <= idx["season"] {
			continue
		}
		dt, _ := ParseDate(row[idx["datetime"]])
		m := Match{
			DateTime:    dt,
			HomeTeam:    strings.TrimSpace(row[idx["home_team"]]),
			AwayTeam:    strings.TrimSpace(row[idx["away_team"]]),
			HomeGoals:   parseInt(row[idx["home_goal"]]),
			AwayGoals:   parseInt(row[idx["away_goal"]]),
			Season:      parseInt(row[idx["season"]]),
			Round:       strings.TrimSpace(row[idx["round"]]),
			Competition: CompBrasileirao,
		}
		matches = append(matches, m)
	}
	return matches, nil
}

// LoadCopa loads Brazilian_Cup_Matches.csv
// Headers: "round","datetime","home_team","away_team","home_goal","away_goal","season"
func LoadCopa(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	r := csv.NewReader(f)
	r.LazyQuotes = true
	headers, err := r.Read()
	if err != nil {
		return nil, err
	}
	idx := indexMap(headers)

	var matches []Match
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Printf("LoadCopa: skipping row due to error: %v", err)
			continue
		}
		if len(row) <= idx["season"] {
			continue
		}
		dt, _ := ParseDate(row[idx["datetime"]])
		m := Match{
			DateTime:    dt,
			HomeTeam:    strings.TrimSpace(row[idx["home_team"]]),
			AwayTeam:    strings.TrimSpace(row[idx["away_team"]]),
			HomeGoals:   parseInt(row[idx["home_goal"]]),
			AwayGoals:   parseInt(row[idx["away_goal"]]),
			Season:      parseInt(row[idx["season"]]),
			Round:       strings.TrimSpace(row[idx["round"]]),
			Competition: CompCopa,
		}
		matches = append(matches, m)
	}
	return matches, nil
}

// LoadLibertadores loads Libertadores_Matches.csv
// Headers: "datetime","home_team","away_team","home_goal","away_goal","season","stage"
func LoadLibertadores(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	r := csv.NewReader(f)
	r.LazyQuotes = true
	headers, err := r.Read()
	if err != nil {
		return nil, err
	}
	idx := indexMap(headers)

	var matches []Match
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Printf("LoadLibertadores: skipping row due to error: %v", err)
			continue
		}
		if len(row) <= idx["season"] {
			continue
		}
		dt, _ := ParseDate(row[idx["datetime"]])
		m := Match{
			DateTime:    dt,
			HomeTeam:    strings.TrimSpace(row[idx["home_team"]]),
			AwayTeam:    strings.TrimSpace(row[idx["away_team"]]),
			HomeGoals:   parseInt(row[idx["home_goal"]]),
			AwayGoals:   parseInt(row[idx["away_goal"]]),
			Season:      parseInt(row[idx["season"]]),
			Stage:       strings.TrimSpace(row[idx["stage"]]),
			Competition: CompLibertadores,
		}
		matches = append(matches, m)
	}
	return matches, nil
}

// LoadBRFootball loads BR-Football-Dataset.csv
// Headers: tournament,home,home_goal,away_goal,away,home_corner,away_corner,home_attack,away_attack,home_shots,away_shots,time,date,...
func LoadBRFootball(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	r := csv.NewReader(f)
	r.LazyQuotes = true
	headers, err := r.Read()
	if err != nil {
		return nil, err
	}
	idx := indexMap(headers)

	var matches []Match
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Printf("LoadBRFootball: skipping row due to error: %v", err)
			continue
		}
		dateCol, ok := idx["date"]
		if !ok || len(row) <= dateCol {
			continue
		}
		dt, _ := ParseDate(row[idx["date"]])
		m := Match{
			DateTime:    dt,
			HomeTeam:    strings.TrimSpace(row[idx["home"]]),
			AwayTeam:    strings.TrimSpace(row[idx["away"]]),
			HomeGoals:   parseInt(row[idx["home_goal"]]),
			AwayGoals:   parseInt(row[idx["away_goal"]]),
			HomeCorner:  parseFloat(row[idx["home_corner"]]),
			AwayCorner:  parseFloat(row[idx["away_corner"]]),
			HomeAttack:  parseFloat(row[idx["home_attack"]]),
			AwayAttack:  parseFloat(row[idx["away_attack"]]),
			HomeShots:   parseFloat(row[idx["home_shots"]]),
			AwayShots:   parseFloat(row[idx["away_shots"]]),
			Tournament:  strings.TrimSpace(row[idx["tournament"]]),
			Competition: CompBRFootball,
		}
		matches = append(matches, m)
	}
	return matches, nil
}

// LoadHistorico loads novo_campeonato_brasileiro.csv
// Headers: ID,Data,Ano,Rodada,Equipe_mandante,Equipe_visitante,Gols_mandante,Gols_visitante,Mandante_UF,Visitante_UF,Vencedor,Arena,OBS
func LoadHistorico(path string) ([]Match, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	r := csv.NewReader(f)
	r.LazyQuotes = true
	headers, err := r.Read()
	if err != nil {
		return nil, err
	}
	idx := indexMap(headers)

	var matches []Match
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Printf("LoadHistorico: skipping row due to error: %v", err)
			continue
		}
		if len(row) <= idx["Ano"] {
			continue
		}
		dt, _ := ParseDate(row[idx["Data"]])
		m := Match{
			DateTime:    dt,
			HomeTeam:    strings.TrimSpace(row[idx["Equipe_mandante"]]),
			AwayTeam:    strings.TrimSpace(row[idx["Equipe_visitante"]]),
			HomeGoals:   parseInt(row[idx["Gols_mandante"]]),
			AwayGoals:   parseInt(row[idx["Gols_visitante"]]),
			Season:      parseInt(row[idx["Ano"]]),
			Round:       strings.TrimSpace(row[idx["Rodada"]]),
			Competition: CompHistorico,
		}
		matches = append(matches, m)
	}
	return matches, nil
}

// LoadFIFA loads fifa_data.csv (has BOM)
// Headers (0-indexed): 0=BOM/"", 1=ID, 2=Name, 3=Age, 4=Photo, 5=Nationality, 6=Flag,
//
//	7=Overall, 8=Potential, 9=Club, 10=Club Logo, 11=Value, 12=Wage, ...
//	21=Position, 22=Jersey Number, 26=Height, 27=Weight
func LoadFIFA(path string) ([]Player, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	r := csv.NewReader(f)
	r.LazyQuotes = true
	r.FieldsPerRecord = -1 // variable fields

	headers, err := r.Read()
	if err != nil {
		return nil, err
	}
	// Strip BOM from first header
	if len(headers) > 0 {
		headers[0] = strings.TrimPrefix(headers[0], "\xef\xbb\xbf")
	}
	idx := indexMap(headers)

	// Build helper to safely get field
	get := func(row []string, key string) string {
		i, ok := idx[key]
		if !ok || i >= len(row) {
			return ""
		}
		return strings.TrimSpace(row[i])
	}

	var players []Player
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Printf("LoadFIFA: skipping row due to error: %v", err)
			continue
		}
		name := get(row, "Name")
		if name == "" {
			continue
		}
		p := Player{
			ID:           parseInt(get(row, "ID")),
			Name:         name,
			Age:          parseInt(get(row, "Age")),
			Nationality:  get(row, "Nationality"),
			Overall:      parseInt(get(row, "Overall")),
			Potential:    parseInt(get(row, "Potential")),
			Club:         get(row, "Club"),
			Position:     get(row, "Position"),
			JerseyNumber: parseInt(get(row, "Jersey Number")),
			Height:       get(row, "Height"),
			Weight:       get(row, "Weight"),
		}
		players = append(players, p)
	}
	return players, nil
}

File: experiments/adrianco/experiment-14/runs/language=typescript_model=opus-4.8-fast_prompt=neutral/rep1/src/loader.ts
/**
 * Brazilian Soccer MCP — Dataset loaders
 * --------------------------------------
 * Context: Reads the six provided Kaggle CSV files from `data/kaggle/` and
 * converts each into the normalized `Match` / `Player` model defined in
 * `types.ts`. Each loader knows the quirks of its source file (column names,
 * language, date format, which competition it represents) and delegates all
 * value-cleaning to `normalize.ts`.
 *
 * Parsing uses `csv-parse` in sync mode with `bom: true` so the FIFA file's
 * leading byte-order-mark and its quoted fields containing commas
 * (e.g. "Jul 1, 2004") are handled correctly.
 */

import { readFileSync } from "node:fs";
import { join } from "node:path";
import { parse } from "csv-parse/sync";
import type { Competition, Match, Player } from "./types.js";
import { canonicalTeam, parseDate, parseInt0, parseNum } from "./normalize.js";

type Row = Record<string, string>;

function read(dataDir: string, file: string): Row[] {
  const text = readFileSync(join(dataDir, file), "utf8");
  return parse(text, {
    columns: true,
    bom: true,
    skip_empty_lines: true,
    relax_column_count: true,
    trim: true,
  }) as Row[];
}

function makeMatch(
  competition: Competition,
  source: string,
  rawHome: string,
  rawAway: string,
  homeGoals: number | null,
  awayGoals: number | null,
  date: string | null,
  season: number | null,
  extra: Partial<Match> = {},
): Match {
  const home = canonicalTeam(rawHome);
  const away = canonicalTeam(rawAway);
  return {
    competition,
    date,
    season,
    round: extra.round ?? null,
    stage: extra.stage ?? null,
    homeTeam: home.display,
    homeTeamId: home.id,
    awayTeam: away.display,
    awayTeamId: away.id,
    homeGoals,
    awayGoals,
    venue: extra.venue ?? null,
    source,
    stats: extra.stats,
  };
}

/** Brasileirao_Matches.csv — Série A 2012–2022. */
export function loadBrasileirao(dataDir: string): Match[] {
  const file = "Brasileirao_Matches.csv";
  return read(dataDir, file).map((r) =>
    makeMatch(
      "Brasileirão Série A",
      file,
      r.home_team,
      r.away_team,
      parseInt0(r.home_goal),
      parseInt0(r.away_goal),
      parseDate(r.datetime),
      parseInt0(r.season),
      { round: r.round || null },
    ),
  );
}

/** Brazilian_Cup_Matches.csv — Copa do Brasil. */
export function loadCopaDoBrasil(dataDir: string): Match[] {
  const file = "Brazilian_Cup_Matches.csv";
  return read(dataDir, file).map((r) =>
    makeMatch(
      "Copa do Brasil",
      file,
      r.home_team,
      r.away_team,
      parseInt0(r.home_goal),
      parseInt0(r.away_goal),
      parseDate(r.datetime),
      parseInt0(r.season),
      { round: r.round || null },
    ),
  );
}

/** Libertadores_Matches.csv — Copa Libertadores. */
export function loadLibertadores(dataDir: string): Match[] {
  const file = "Libertadores_Matches.csv";
  return read(dataDir, file).map((r) =>
    makeMatch(
      "Copa Libertadores",
      file,
      r.home_team,
      r.away_team,
      parseInt0(r.home_goal),
      parseInt0(r.away_goal),
      parseDate(r.datetime),
      parseInt0(r.season),
      { stage: r.stage || null },
    ),
  );
}

/** novo_campeonato_brasileiro.csv — Série A 2003–2019 (Portuguese columns). */
export function loadNovoBrasileirao(dataDir: string): Match[] {
  const file = "novo_campeonato_brasileiro.csv";
  return read(dataDir, file).map((r) =>
    makeMatch(
      "Brasileirão Série A",
      file,
      r.Equipe_mandante,
      r.Equipe_visitante,
      parseInt0(r.Gols_mandante),
      parseInt0(r.Gols_visitante),
      parseDate(r.Data),
      parseInt0(r.Ano),
      { round: r.Rodada || null, venue: r.Arena || null },
    ),
  );
}

const TOURNAMENT_MAP: Record<string, Competition> = {
  "Serie A": "Brasileirão Série A",
  "Serie B": "Brasileirão Série B",
  "Serie C": "Brasileirão Série C",
  "Copa do Brasil": "Copa do Brasil",
};

/** BR-Football-Dataset.csv — extended stats across Série A/B/C & Copa. */
export function loadBrFootball(dataDir: string): Match[] {
  const file = "BR-Football-Dataset.csv";
  const out: Match[] = [];
  for (const r of read(dataDir, file)) {
    const competition = TOURNAMENT_MAP[r.tournament];
    if (!competition) continue;
    const date = parseDate(r.date);
    // BR-Football has no season column, so we infer it from the date's year.
    // Brazilian league seasons run within a calendar year (≈May–Dec) — the sole
    // modern exception is the COVID-delayed 2020 season, which finished in
    // Feb 2021. So Série A/B/C matches played in Jan/Feb belong to the *previous*
    // year's season; without this they'd be mis-bucketed and fail to dedupe
    // against the authoritative, explicitly-seasoned datasets.
    let season = date ? Number(date.slice(0, 4)) : null;
    if (season !== null && date) {
      const month = Number(date.slice(5, 7));
      if (month <= 2) season -= 1;
    }
    out.push(
      makeMatch(
        competition,
        file,
        r.home,
        r.away,
        parseInt0(r.home_goal),
        parseInt0(r.away_goal),
        date,
        season,
        {
          stats: {
            homeCorners: parseNum(r.home_corner),
            awayCorners: parseNum(r.away_corner),
            homeShots: parseNum(r.home_shots),
            awayShots: parseNum(r.away_shots),
            homeAttacks: parseNum(r.home_attack),
            awayAttacks: parseNum(r.away_attack),
            totalCorners: parseNum(r.total_corners),
          },
        },
      ),
    );
  }
  return out;
}

/** fifa_data.csv — FIFA player database. */
export function loadPlayers(dataDir: string): Player[] {
  const file = "fifa_data.csv";
  return read(dataDir, file)
    .filter((r) => r.Name && r.ID)
    .map((r) => {
      const club = canonicalTeam(r.Club || "");
      return {
        id: parseInt0(r.ID) ?? 0,
        name: r.Name,
        age: parseInt0(r.Age),
        nationality: r.Nationality || "",
        overall: parseInt0(r.Overall),
        potential: parseInt0(r.Potential),
        club: r.Club || "",
        clubId: r.Club ? club.id : "",
        position: r.Position || "",
        jerseyNumber: parseInt0(r["Jersey Number"]),
        height: r.Height || "",
        weight: r.Weight || "",
        preferredFoot: r["Preferred Foot"] || "",
        value: r.Value || "",
        wage: r.Wage || "",
      } satisfies Player;
    });
}

export interface LoadedData {
  matches: Match[];
  players: Player[];
}

/** "Richness" score used to choose the best record among duplicates. */
function richness(m: Match): number {
  let s = 0;
  if (m.stats) s += 8;
  if (m.round) s += 2;
  if (m.venue) s += 1;
  if (m.stage) s += 1;
  return s;
}

/** Merge two records of the same real-world match, keeping the most data. */
function mergeMatch(a: Match, b: Match): Match {
  const [rich, poor] = richness(a) >= richness(b) ? [a, b] : [b, a];
  // Sources disagree on dates by ~1 day (local vs UTC kickoff). Prefer the
  // earlier date, which corresponds to the Brazilian local match day.
  const dates = [a.date, b.date].filter((d): d is string => !!d).sort();
  return {
    ...rich,
    date: dates[0] ?? rich.date ?? poor.date,
    round: rich.round ?? poor.round,
    stage: rich.stage ?? poor.stage,
    venue: rich.venue ?? poor.venue,
    homeGoals: rich.homeGoals ?? poor.homeGoals,
    awayGoals: rich.awayGoals ?? poor.awayGoals,
    stats: rich.stats ?? poor.stats,
  };
}

/**
 * Collapse duplicate matches that appear across overlapping datasets. The same
 * Série A games are present in up to three files (Brasileirao_Matches,
 * novo_campeonato_brasileiro, BR-Football-Dataset); without this, seasons in the
 * 2012–2019 overlap would be double/triple-counted in standings and stats.
 *
 * Keying strategy: within a single competition + season, an ordered (home, away)
 * fixture is unique (round-robin leagues play each ordered pairing once; cup
 * legs swap home/away). We therefore key on competition + season + home id +
 * away id. This is deliberately date-independent because the sources disagree on
 * dates by up to a day (local kickoff vs UTC). When season is unknown we fall
 * back to a date-based key; records with neither are kept as-is.
 */
export function dedupeMatches(matches: Match[]): Match[] {
  const byKey = new Map<string, Match>();
  const kept: Match[] = [];
  for (const m of matches) {
    let key: string | null = null;
    if (m.season !== null) {
      key = `${m.competition}|S${m.season}|${m.homeTeamId}|${m.awayTeamId}`;
    } else if (m.date) {
      key = `${m.competition}|${m.date}|${m.homeTeamId}|${m.awayTeamId}`;
    }
    if (key === null) {
      kept.push(m);
      continue;
    }
    const existing = byKey.get(key);
    byKey.set(key, existing ? mergeMatch(existing, m) : m);
  }
  return [...byKey.values(), ...kept];
}

/** Load and merge every dataset into a single, de-duplicated collection. */
export function loadAll(dataDir: string): LoadedData {
  const matches = dedupeMatches([
    ...loadBrasileirao(dataDir),
    ...loadNovoBrasileirao(dataDir),
    ...loadCopaDoBrasil(dataDir),
    ...loadLibertadores(dataDir),
    ...loadBrFootball(dataDir),
  ]);
  const players = loadPlayers(dataDir);
  return { matches, players };
}
